### DISCUSSION TITLE
Strongswan mysql / mariadb configuration

### DISCUSSION THREAD

**Discussion by dotmitsu** [strongswan/strongswan#1742] [Q&A] [answered]
Hi all. I am trying to get strongswan working with mariadb. I use this plugin https://docs.strongswan.org/docs/5.9/plugins/sql.html and mysql plugin. Plugins turned on. I created a database structure using this https://github.com/strongswan/strongswan/blob/master/src/pool/mysql.sql but I can't find any example of how to fill in the tables correctly. Can you help me please? Can someone share a dump of their database or give a link to information about it?

**Comment by tobiasbrunner** [comment]
On the first page you linked, there is a link to more information about some of the database fields (see [here](https://docs.strongswan.org/docs/5.9/config/sqliteDbSchema.html), applies similarly to MySQL). And there links to example data files that are used in our testing environment.

**Comment by dotmitsu** [comment]
Thanks for unswer) I saw it, but didn't understand how to fill in the fields.
![изображение](https://github.com/strongswan/strongswan/assets/42657211/2c9629f6-01c4-4500-b1f7-ca12177e0a8f)
What does this data mean X'c0a80001'?
Is it an encrypted value?

**Reply by tobiasbrunner** [reply]
No, just a hexadecimal encoding of the IP address (192 = 0xc0, 168 = 0xa8, 0 = 0x00, 1 = 0x01). You can use the script mentioned there to convert any kind of identity to their binary encoding usable in SQL.

**Reply by dotmitsu** [reply]
Can I store ip adress as string? Without encoding?

**Reply by dotmitsu** [reply]
And I have another question. I turned on sql and mysql plugin. Do I need to specify somewhere to take the configuration from the database, and not from the ipsec.secrets and ipsec.conf (strongswan.conf) files?

**Reply by tobiasbrunner** [reply]
> Can I store ip adress as string? Without encoding?

No, only binary encoding of the identities is currently supported.



> Do I need to specify somewhere to take the configuration from the database, and not from the ipsec.secrets and ipsec.conf (strongswan.conf) files?

You only have to specify the database URI via `charon.plugins.sql.database`. Then configs may be provided via that database.

**Comment by dotmitsu** [comment]
Thanks for the replies. I decided to abandon the idea of uding this plugin because there is no normal instruction describing how to fill in the fields in mariadb. There are a lot of fields with the not Null parameter that are not clear how to fill in. Unfortunately the documentation for this plugin is very poor.

**Reply by tobiasbrunner** [reply]
> because there is no normal instruction describing how to fill in the fields

What do you mean with "normal instruction"? What exactly did you expect?

> There are a lot of fields with the not Null parameter that are not clear how to fill in.

Many have default values, and they pretty much correspond to settings you also find in swanctl.conf. And again, there are detailed information about some of the fields, scripts to generate the encoding of identities and keyids (and binary blobs like certificates), and lots of examples.  Which fields were unclear to you?

**Comment by dotmitsu** [comment]
Thanks for your reply) I had difficulty understanding the fields, what to fill in where. I imported the database from github and got this structure. 
![picture1](https://github.com/strongswan/strongswan/assets/42657211/f1b295e6-62d3-41ac-98af-ab845bba5bc5)
This is my ipsec.conf file, which works good, But I cant understand how I can fill this fields in database. 
```
config setup
    charondebug="ike 1, knl 1, cfg 0"
    uniqueids=never

conn ikev2-vpn
    auto=add
    compress=no
    type=tunnel
    keyexchange=ikev2
    fragmentation=yes
    forceencaps=yes
    dpdaction=clear
    dpddelay=300s
    rekey=no
    left=%any
    leftid=77.222.43.35
    leftcert=server-cert.pem
    leftsendcert=always
    leftsubnet=0.0.0.0/0
    right=%any
    rightid=%any
    rightauth=eap-mschapv2
    rightsourceip=172.27.27.0/24
    rightdns=8.8.8.8,8.8.4.4
    rightsendcert=never
    eap_identity=%identity
    ike=aes256gcm16-aes256gcm12-aes128gcm16-aes128gcm12-sha256-sha1-modp2048-modp4096-modp1024,aes256-aes128-sha256-sha1-modp2048-modp4096-modp1024,3des-sha1-modp1024!
    esp=aes128gcm12-aes128gcm16-aes256gcm12-aes256gcm16-modp2048-modp4096-modp1024,aes128-aes256-sha1-sha256-modp2048-modp4096-modp1024,aes128-sha1-modp2048,aes128-sha1-modp1024,3des-sha1-modp1024,aes128-aes256-sha1-sha256,aes128-sha1,3des-sha1!
```

ipsec.secrets
```
: RSA "server-key.pem"
user1_vpn : EAP "Password1"
user2_vpn : EAP "Password2"
```

**Reply by dotmitsu** [reply]
I found this file https://www.strongswan.org/testing/testresults/sql/rw-eap-aka-rsa/moon.ipsec.sql but can't execute it in mariadb because of syntax error

**Reply by tobiasbrunner** [reply]
> I had difficulty understanding the fields, what to fill in where.

Which ones exactly? 

> This is my ipsec.conf file, which works good, But I cant understand how I need to fill this fields in database.

Might have been easier if you first converted that to swanctl.conf (see e.g. [here](https://wiki.strongswan.org/projects/strongswan/wiki/fromipsecconf)) as that makes it a bit clearer that some settings apply to the IKE_SA, some to the authentication (peer) and some to the CHILD_SA.


> ipsec.secrets

Note that if your goal was to store these client credentials in the database, it's perfectly fine to do so in combination with a config in ipsec.conf/swanctl.conf. The config doesn't have to be in the SQL database for that.



> but can't execute it in mariadb because of syntax error

We use SQLite in the testing environment, so that might not work. But just copying that will probably not be what you want. Just learn from it and adapt it to your use case.

**Reply by dotmitsu** [reply]
Which tables I need to fill if I need store only user login and password? 
And one more question a bit off topic. Maybe stupid question. What type of user authentication can I use so that they do not need to copy ca-cert.pem to the users device? So that they can simply connect with a username and password. I would be grateful for an answer, because the manual does not describe such moments.

**Reply by dotmitsu** [reply]
I want to store user credentials in 1 database and 3 VPN servers get user credentiials from this database.

**Reply by tobiasbrunner** [reply]
> Which tables I need to fill if I need store only user login and password?

`identities` stores the usernames, `shared_secrets` the passwords and `shared_secret_identity` connects the two (because secrets can be associated with multiple identities).





> What type of user authentication can I use so that they do not need to copy ca-cert.pem to the users device?

Only pre-shared keys (PSK) won't require a certificate for the server. But many clients won't support that. Also, if you get a server certificate from a CA the clients already trust (e.g. Let's Encrypt) you don't have to install any certificates on the client.

**Reply by tobiasbrunner** [reply]
> I want to store user credentials in 1 database and 3 VPN servers get user credentiials from this database.

I assume the same ones? Not different ones for each server?

**Reply by dotmitsu** [reply]
Yes, credentials will be the same

**Reply by dotmitsu** [reply]
Thanks a lot, you helped me a lot.

Sorry for so many questions in 1 thread. I can create separate ones if needed. There are 2 more questions.
1) How to limit the number of possible connections from 1 account? For example, no more than 5 connections at the same time.
2) Does strongswan have functionality to limit the speed for each client (For example 10Mbps) or is it necessary to do it using iptables?

**Reply by tobiasbrunner** [reply]
> Is there an instruction somewhere on how to generate keys for Stringswan using lets encrypt? I've generated many times for nginx and the like, but never for strongswan. Is this key appropriate for the type of EAP authentication I'm using?

Since you just generate a server certificate (and private key) this has nothing directly to do with EAP. And the procedure is basically the same as for web server certificates. See https://github.com/strongswan/strongswan/discussions/1410#discussioncomment-4109873 for some hints on how to deploy this when using certbot (uses swanctl but works similarly with the old config backend).


> 1. How to limit the number of possible connections from 1 account? For example, no more than 5 connections at the same time.

There is currently no such mechanism built into strongSwan. You could implement something like that yourself (e.g. custom plugin or a script in the [ext-auth plugin](https://docs.strongswan.org/docs/5.9/plugins/ext-auth.html)) but unless you do that via shared storage (e.g. a database) yourself, that will be per VPN server. However, since you intend to use the same authentication database with multiple servers anyway, you maybe want to look into RADIUS ([eap-radius plugin](https://docs.strongswan.org/docs/5.9/plugins/eap-radius.html)), e.g. FreeRADIUS, which supports such restrictions even when clients authenticate from different VPN servers.

> 2\. Does strongswan have functionality to limit the speed for each client (For example 10Mbps) or is it necessary to do it using iptables?

No. You have to configure that yourself e.g. via `tc` and `iptables` from an updown/vici script (it might be possible to configure the rules globally so they apply per IP address from the virtual IP pool, but not sure).

**Reply by dotmitsu** [reply]
Thanks for all answers))

**Comment by dotmitsu** [comment]
Hello again! I've installed StrongSwan on Ubuntu 22.04 and also separately compiled StrongSwan on my local machine. I extracted two files libstrongswan-mysql.so and libstrongswan-sql.so. After that, I copied them to the server in the directory /usr/lib/ipsec/plugins/.
Next, I created two files: /etc/strongswan.d/charon/mysql.conf and /etc/strongswan.d/charon/sql.conf. In the StrongSwan logs, I observed a successful connection to MariaDB, but the required tables were missing. I downloaded the mysql.sql file from [GitHub](https://github.com/strongswan/strongswan/blob/master/src/pool/mysql.sql) and restored the database.

Now there are no errors. swanctl --stats shows mysql and sql are loaded. But I have a problem. I created an entry in the identities table 1, 2, X'7465737475736572'
![изображение](https://github.com/strongswan/strongswan/assets/42657211/56cc6fbc-1058-4c86-a4b5-105c6adb4604)

then in shared_secrets 1, 2, X'7465737470617373776f7264'.
![изображение](https://github.com/strongswan/strongswan/assets/42657211/0179f300-8895-4343-87f1-9734b57bb259)

I then linked them in shared_secret_identities with the entry 1, 1.
![изображение](https://github.com/strongswan/strongswan/assets/42657211/ae10e18d-6bef-4528-a568-57e7e61f145d)

It seems that everything should work, but authorization does not pass. I tried to enter other types of identities, but I still can't log in. Maybe I need to remove something from the ipsec.secrets file or, on the contrary, add some parameters so that the data is read from the database?

**Reply by dotmitsu** [reply]
@tobiasbrunner

**Reply by tobiasbrunner** [reply]
The data might not be correct (I don't know what GUI you use there or what the actual output should look like but the data is binary).

**Reply by dotmitsu** [reply]
@tobiasbrunner data is binary. I converted it by using id2sql script. GUI is Navicat software. I fill data by this manual https://docs.strongswan.org/docs/5.9/config/sqliteDbSchema.html

**Reply by tobiasbrunner** [reply]
Maybe not if you pasted the console output to the GUI. Did you actually add the entries with an SQL script?

**Reply by dotmitsu** [reply]
I wrote the commands through sql and now the login and password are displayed normally)), but still does not allow to log in yet
![изображение](https://github.com/strongswan/strongswan/assets/42657211/3638e1ab-085d-4208-a0e0-80882974eeaf)
![изображение](https://github.com/strongswan/strongswan/assets/42657211/3e3acb07-ffea-4765-97f7-92ae30c4b429)
![изображение](https://github.com/strongswan/strongswan/assets/42657211/133f4ffd-5510-4333-8373-9c22a18c991f)

**Reply by dotmitsu** [reply]
@tobiasbrunner

**Reply by dotmitsu** [reply]
I switched mariadb to debug mode. When I am connecting to vpn, I see this requests. I tested this request in my DB. It doesn't return anything.

**Reply by dotmitsu** [reply]
It works. I didn't understand why. I looked at the query that goes to the database and saw that there should be a record in the identites table with type 2 and the address of my server. I added such record there and made the necessary record in shared_secret_identity and the connection is now working? Why is such record needed? I don't understand.
I am using certificates generated by letsencrypt

**Reply by dotmitsu** [reply]
Strongswan uses this query
![изображение](https://github.com/strongswan/strongswan/assets/42657211/ce68a179-df74-4543-90bc-2c3eb06d977d)

**Reply by dotmitsu** [reply]
I don't understand why we have `AND m.data = 'ADDRESS_OF_SERVER'` in query from strongswan. Didn't work exactly because of this condition

**Reply by tobiasbrunner** [reply, marked as answer]
> Why is such record needed? I don't understand.

PSKs are currently matched on both identities, client and server.

**Reply by dotmitsu** [reply]
In the current implementation, the user can type in the VPN client instead of the user's login the server address and his password and log in. We have slightly changed the structure of the database. Added 2 tables. A table with servers and a table for matching servers and users. And changed 1 sql query in sql_cred.c file. Now, the fields of the table are linked using foreign keys. Now, when a user is deleted, all his records in the tables associated with him will be automatically deleted. To leave no trash.  If you want, I can create a patch, check it out and if you like the realization, you can accept it into the main branch. I think our implementation is more correct. In it, you can specify the servers on which the user can log in, now the user can log in only with his login as before, but cannot log in by entering the server address instead of his login.
And we can write a manual for this. But only for user creds in mysql/mariadb. Because we doesn't use server configurations in mysql.

**Reply by tobiasbrunner** [reply]
I don't think more tables are necessary (you can of course do that on your end to manage server-specific identities). But maybe see the discussion in #184 on changing the query to make sure the client and server identities are not the same.

**Comment by dotmitsu** [comment]
I edited the sql plugin for my tasks, now it works as I need. And then in the standard implementation, if you add the server address to the table for each user (so that authorization works), then any user will be able to enter the server address instead of his login and authorization will occur.

**Comment by dotmitsu** [comment]
@tobiasbrunner I have one more problem. How can I make that strongswan reread creds or identities from mysql/mariadb when user is already connected to strongswan vpn? I removed the user from identities, but the user is still online until he disconnects and tries to reconnect. Does strongswan check at some interval whether the user can connect to the vpn or not?

**Reply by tobiasbrunner** [reply]
No, why should it? The database stores credentials for authentication, nothing else.

**Reply by dotmitsu** [reply]
Hello. I have an additional question. Does mysql plugin have ssl connection support?

**Reply by tobiasbrunner** [reply]
In what way? To connect to the database via TLS? If so, I guess that depends on the options with which MySQL/MariaDB and the corresponding client library was compiled and how the database was configured (whether it allows or even enforces encrypted connections). We currently don't set any TLS-related options, so the default of e.g. ` SSL_MODE_PREFERRED` should apply when the connection is established, but that doesn't enforce any specific CA certificate or identity (that would require changes to the plugin e.g. to configure the CA certificate and enforce a specific mode). Feel free to open a feature request.

**Reply by dotmitsu** [reply]
We use require-secure-transport = on in Mariadb. Every other app connects normally to db, but strongswan can't authenticate. In mariadb client we need to use --ssl parameter for successfull connection. Can we use something like this for strongswan mysql plugin?

**Reply by tobiasbrunner** [reply]
As I said, the _mysql_ plugin does not set any TLS-specific options, so the defaults will apply (which might be different between MySQL and MariaDB actually, as I didn't see `MYSQL_OPT_SSL_MODE` in the MariaDB docs). If that doesn't work, code changes are necessary.

> Every other app connects normally to db, but strongswan can't authenticate.

What does that mean? Do you require clients to authenticate with a certificate?
